platform/bare_metal/mbed_lib.json

```diff
@@ -1,4 +1,4 @@
 {
   "name": "bare-metal",
-  "requires": ["platform", "drivers"]
+  "requires": ["platform", "drivers", "minimal-printf"]
```

Comment:
Does it mean it's default in bare-metal mode?

Reply:
That just includes it in the build. I'm not convinced it needs to be added to the requires list. You can require it if you decide to activate it.

This might be another case for adding more places you can put the `requires` directive. Some work has shown that targets might need to do this - eg Nordic boards refer to some code with `mbed_lib.json`, which means you currently need `requires: "nordic"` if using this mode with those boards. That should be in targets.json.

Maybe this `requires` should be in the mprintf.json profile.

Reply:
Alternatively, the minimal printf may fit inside platform anyway. What it's doing is similar to what I'm doing in #11039, providing a C/C++ library alternative. All such mechanisms are currently inside `platform`.

This isn't really a "feature" - it doesn't /add/ anything.

Maybe `platform/csupport/minimal-printf`?

Reply:
It was a suggestion from @bulislaw to put minimal-printf under feature. I agree that it is not really a feature so I can move to `platfom/minimal-printf`(I prefer a shorter path) if Bartek is okay.

As for the orginal question about "requires", I don't know much about this feature. The code change was originally made by @ithinuel. I don't think it means that this is the default mode for bare-metal since minimal-printf is activated by invoking the mprintf profile `-- profile mprintf`

@kjbracey-arm what do you mean by this `requires` should be in the mprintf.json profile?

Reply:
`requires` things should go where they're needed, as a tree. So if the application requires `x`, it has "requires: x", and if x requires y, its own mbed_lib.json should have "requires: y".

This should be pretty transparent to the application, and there shouldn't be a case where you have to have an "indirect" requires statement. The application shouldn't need to say "requires: y" if it's only directly using "x".

Here, "bare-metal requires minimal-printf" is not true - it doesn't, therefore the requires should not be there.

"application requires minimal-printf" may be true,  but only if it's trying to enable the minimal-printf options. So it would be reasonable to require the application to have the requires, if it has minimal-printf options in its mbed_app.json.

But as the application isn't doing that directly, but via the `--profile`, so it would make sense for the profile itself to have the requires. The mbed_app.json doesn't know whether it requires it or not - it depends on the --profile selected on the CLI.

And that's similar to the problem we've got of "requires" depending on the target selected on the CLI. There needs to be a way to get from CLI option -> target-or-profile -> additional require, without either the app needing to mess with its JSON, or having to stick a dummy requires in like this one to make it "always on".

Just sticking it in "platform" with its options in platform/mbed_lib.json would side-step the issue, but if it retains its own platform/minimal-printf/mbed_lib.json, then the issue remains.

I don't think it's actually possible in the tooling to have targets or profiles adding "requires" dependencies at present though.

The issue for this lib is pretty minor - it's small enough to not be a big deal having it always there in the build, but we do need an answer for the big CLI-option-dependent things like "requires: nordic".

Reply:
> Just sticking it in "platform" with its options in platform/mbed_lib.json would side-step the issue, but if it retains its own platform/minimal-printf/mbed_lib.json, then the issue remains.

That is a simple and good solution. I'll go ahead with that. Thanks Kevin

Reply:
As per discussion:
- requires `minimal-printf` removed from `platform\bare_metal\mbed_lib.json`
- `platform\minimal-printf\mbed_lib.json` removed